### PR TITLE
Blog authorization

### PR DESCRIPTION
**Pull request by Estherstarr**
In this PR, I implemented the following
- Install CanCanCan in your project.
- Add a `role` column to the users table. Remember to use a migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column `role` has value `"admin"`). Use CanCanCan for this authorization.
    - For that you need to implement the post deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column `role` has value `"admin"`). Use CanCanCan for this authorization.
    - For that you need to implement the comment deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

### CODE REVIEW

**Review by omar-muhamad — COMMENTED** [review]
# Invalid ❌

Hi @Estherstarr and @Elijahdre 

- [ ] This is a peer-programming project milestone and I can see only one team member has committed all changes, So please make sure that all team members have the same amount of work shared and committed to GitHub.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **Does not count** into the code reviews limit._

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    can :read, Post, public: true
+
+    can :read, Comment, public: true
+
+    return unless user.present?
+
```
Comment (Davidosky007) [inline app/models/ability.rb]: * Kindly ensure
>   A user can delete a post if it is theirs or if they have an admin role
> A user can delete a comment if it is theirs or if they have an admin role
* As currently a user can delete posts and comments not created by them Kindly fix this 🙏 below is a recording of the bug

https://user-images.githubusercontent.com/52497246/216737871-549ee859-73e2-459f-9a57-8821c546877d.mp4

**Review by Davidosky007 — CHANGES_REQUESTED** [review]
Hi @Estherstarr @Elijahdre 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
## Highlights

* No linter errors 👍 
* Well written PR description 👍 
* Good looking readme file 👍 
## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @davidsoky007 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/posts/index.html.erb [hunk 2]
```diff
@@ -19,13 +19,13 @@
...
-         method: :delete do %>
-            <button class = "bg-red-600	mt-3 hover:bg-red-700 text-white
-             font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline" >Delete</button>
-        <% end %>
-
+          <% if can?(:destroy, post) %>
+            <%= button_to  user_post_path(@user.id, post.id), method: :delete do %>
+              <button class="bg-red-600	mt-3 hover:bg-red-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline">
+                Delete
+              </button>
```
Comment (elisha2kyakpo1) [inline app/views/posts/index.html.erb:26]: - [ ] Interestingly this button increments the likes on a post. We would want to delete a post if I am a user who created it. Kindly fix that.

File: app/models/ability.rb [hunk 3]
```diff
@@ -3,20 +3,17 @@ class Ability
...
-
     can :read, Comment, public: true
 
     return unless user.present?
 
-    can %i[read destroy], Post, user: user
-
-    can %i[read destroy], Comment, user: user
+    can %i[read destroy], Post, author: user
+    can %i[read destroy], Comment, author: user
```
Comment (elisha2kyakpo1) [inline app/models/ability.rb]: - [ ] Kindly fix the issue of deleting other users' posts and comments as requested by the previous reviewer. I created two accounts, and I created posts and comments in both, unfortunately, I was able to delete a comment for another user.

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
Hi 👋 @Estherstarr  and @Elijahdre, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.


 ## Highlights

👍 You did a great job making the requested changes.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Deepakdanger** [comment]
Hi your person.rb should be like this

class Person < Nameable

  attr_reader :id
  attr_accessor :name, :age ,:rental
  def initialize(name, age = 'Unknown', parent_permission: true)
    ........
    @rental = []
  end
.......
end

**Review by EricMbouwe — APPROVED** [review]
Hi @Estherstarr, 

Your project is complete! ⚡🔥  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 🎉🔥
- Requested changes ✔
- No linter errors ✔

## Optional suggestions ☂

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**